clippy_lints/src/dereference.rs

```diff
@@ -353,8 +425,321 @@ fn is_linted_explicit_deref_position(parent: Option<Node<'_>>, child_id: HirId)
     }
 }
 
+// Checks if the expression for the given id occurs in a position which auto dereferencing applies.
+// Note that the target type must not be inferred in a way that may cause auto-deref to select a
+// different type, nor may the position be the result of a macro expansion.
+//
+// e.g. the following should not linted
+// macro_rules! foo { ($e:expr) => { let x: &str = $e; }}
+// foo!(&*String::new());
+// fn foo<T>(_: &T) {}
+// foo(&*String::new())
+fn is_stable_auto_deref_position(tcx: TyCtxt<'tcx>, typeck: &'tcx TypeckResults<'_>, id: HirId) -> bool {
```

Comment:
Hmm. I think it would be better to do this the other way around. That is, in `check_expr`, look for a "stable auto deref position", and then proceed inward to check for explicit deref. This way you don't have to climb the HIR with `parent_iter`.

Reply:
Walking down would require implementing `check_local`, `check_item`, `check_impl_item` and `check_trait_item` in addition to `check_expr`. It also requires walking from a block to all the `break` expressions that target it, which I'm not sure how to do other than visiting everything.

Then there's the interaction with `explicit_deref_methods` which can occur in other positions. `explicit_auto_deref` should lint instead when possible, rather than both lints triggering.

Reply:
Okay I'm convinced. :)

One other idea. Use `tcx.hir().opt_span()` to check `in_macro` in one place, perhaps after `match parent {}`.

Reply:
Actually I think it's good for both lints to trigger. The user may disagree with one of the lints but agree with the other.

Reply:
If they disagree with a lint they can just `#[allow]` it. Then the other lint can trigger instead.

In this case both lints would trigger on the same code with conflicting suggestions, which breaks rustfix.

Reply:
It is valid to use Clippy without adding `#[allow]` on undesired lint cases. I don't think it would break rustfix as of rust-lang/cargo#5842.

Reply:
In this case both the code and the suggestion for `explicit_deref_methods`  (`x.deref()` -> `&*x`) trigger `explicit_auto_deref` (would change both to `&x`), and this will always be the case. If someone doesn't want to apply `explicit_auto_deref ` they would have to allow it anyways. If they do want to apply it, it's the one that would take priority.

Reply:
> they would have to allow it anyways

not if they are not using rustfix.

I see your point that the suggestions are redundant. But there is a risk that the user will ignore `explicit_auto_deref` because they want the deref to be explicit, and at the same time, they *would* be willing to change to `*`, but that suggestion is suppressed. This risk outweighs the redundancy IMO. You can display the preferred lint first so that the user won't fix it twice. Also displaying both lints is more informative for someone learning Rust since there are two learning opportunities from the same code.

Anyways, that's just my opinion and this is a small corner case.

Reply:
I guess they could just leave clippy forever triggering `explicit_auto_deref`. If that's a real use case then both being suggested is reasonable.